WasatchNET/MultiChannelWrapper.cs

```diff
@@ -0,0 +1,407 @@
+        // Public attributes
+        ////////////////////////////////////////////////////////////////////////
+
+        /// <summary>
+        /// A convenience accessor to iterate over valid positions (channels).
+        /// </summary>
+        public SortedSet<int> positions { get; private set; } = new SortedSet<int>();
+
+        ////////////////////////////////////////////////////////////////////////
+        // Private attributes
+        ////////////////////////////////////////////////////////////////////////
+
+        const int TRIGGER_PULSE_WIDTH_MS = 50;
+
+        static MultiChannelWrapper instance = null;
+
+        Driver driver = Driver.getInstance();
+        Dictionary<int, Spectrometer> specByPos;
+        Logger logger = Logger.getInstance();
+
+        Spectrometer specTrigger = null;
+        Spectrometer specFan = null;
+
+        static object mut = new object();
+
+        ////////////////////////////////////////////////////////////////////////
+        // Lifecycle
+        ////////////////////////////////////////////////////////////////////////
+
+        /// <summary>
+        /// Get a handle to the MultiChannelWrapper singleton.
+        /// </summary>
+        public static MultiChannelWrapper getInstance()
+        {
+            lock (mut)
+            {
+                if (instance is null)
+                    instance = new MultiChannelWrapper();
+                return instance;
+            }
+        }
+
+        MultiChannelWrapper()
+        {
+            reset();
+        }
+
+        void reset()
+        {
+            specByPos = new Dictionary<int, Spectrometer>();
+            positions = new SortedSet<int>();
+            specTrigger = null;
+            specFan = null;
+        }
+
+        /// <summary>
+        /// Initialize all connected spectrometers, determine positions in the multi-channel
+        /// system, find out which are configured to provide the external hardware trigger
+        /// and fan control, etc.
+        /// </summary>
+        /// <returns>true on success</returns>
+        public bool open()
+        {
+            reset();
+
+            var count = driver.openAllSpectrometers();
+            logger.info($"found {count} spectrometers");
+            for (var i = 0; i < count; i++)
+            {
+                var spec = driver.getSpectrometer(i);
+                var sn = spec.serialNumber;
+
+                logger.info("found {0} {1} with {2} pixels ({3:f2}, {4:f2})", 
+                    spec.model, sn, spec.pixels, spec.wavelengths[0], spec.wavelengths[spec.pixels-1]);
+
+                ////////////////////////////////////////////////////////////////
+                // Parse EEPROM.userText
+                ////////////////////////////////////////////////////////////////
+
+                var pos = spectrometerCount + 1;
+
+                var attributes = spec.eeprom.userText.Split(';');

```

Comment:
I know it's somewhat implicit, but I feel like a reference to the readme for what's going on here could be useful

Reply:
Gotcha.